### ISSUE TITLE
[RFC][Training] Add dflash support in training

### ISSUE THREAD

**Issue by shanjiaz** [vllm-project/speculators#248]
# Background
DFlash has recently emerged in a [blog post](https://z-lab.ai/projects/dflash/) as a potentially superior method for speculative decoding compared to Eagle-3.  

<img width="617" height="321" alt="Image" src="https://github.com/user-attachments/assets/8948a4cc-77d7-4735-a106-40a96b89f573" />

It uses a 5-layer diffusion-llm to predict a block of tokens as a draft for the base model, where the prediction is conditioned on the hidden states of the base model, similar to Eagle 3.  The architecture is shown above, and a reference implementation of the inference model is provided [here](https://github.com/z-lab/dflash/blob/main/model/dflash.py).  

The blog claims potentially up to 2-3x larger speedups than Eagle 3 for synchronous requests, and inference is actively being implemented in vLLM.

# Training algorithm

The model needs to be trained to take in an input of hidden states from the model, an embedded token, and some mask tokens (which will become the draft) and predict the correct tokens for the draft.  

Training Inputs and Attention Mask: 
For a single document/ coherent sequence:  
Given sequence $(S_1…S_L)$ of  length $L$ and mask tokens $M_1…M_L$, and block size $B$ (also the draft length) with hidden dimension $D$, 
Query: Shape:  $[B,L,D]$ composed of noise (or mask) $M_1…M_L$.  
Key/Value:  Shape: $[B, 2L, D]$ concatenated from $[S_1…S…L]$ and $[M_1…M_L]$
Attention Mask: Shape: $[L, 2L]$  
For example, for a sequence of size 128 and block size 8, it should look like this:  

*As of Thursday afternoon the LMSYS attn block was incorrectly causal in the mask block. 

- Position IDs: ${0…L-1, 0…L-1}$ concat

<img width="640" height="336" alt="Image" src="https://github.com/user-attachments/assets/3083ed1b-867d-417e-a988-3bf55d6ca7ea" />

- Target: Shape: $[L]$ The target ids for the model
- Loss mask: $[L]$ Loss mask (same as for eagle3) which only trains on tokens generated by the base model.  Based on the chat template.  - Generated by Eagle 3 data generation code already.  
- Potential:  T2D and D2T tensors:  Vocabulary matching same as we use for Eagle 3.  I don't see any reason to exclude this functionality from DFlash.  
- Difference from Eagle 3:  We only need one forward pass to train the model, we do not need TTT, so in that respect it is simpler.  

# Required Changes

### Single Forward Pass:

Dflash runs a single forward step which produces a tensor of shape $[vocab\\_size,L]$  This is in contrast to the eagle3 train time test paradigm which runs the drafter forward for $T$ steps.  The drafter produces a block of tokens of length $B$ for each prefix (total of $L/B$ prefixes).  This block structure is entirely accomplished using the attention mask. We calculate the loss using the same function as eagle 3.  We then apply the loss mask to this and calculate gradients on the model.  
 
### Sequence trimming: 

Theoretically we would like to start a block with every possible prefix.  Given a sample $[S_1…S_L]$ of length $[L]$, this means a prefix with $[S_1]$, a prefix with $[S_1, S_2]$, and so on.  However, in order to make the initial pass at training, it is probably easiest to simply clip the samples to size $integer \times B$, and start counting blocks at 0, as shown in the attention mask in the background section.  This is because we have to append mask (noise) tokens to the K and V tensors for every token that we want to predict, and include those in the Query tensor. Instead of running the algorithm forward for train time test steps like in eagle3, and appending the outputs of the model, we want to provide L mask tokens as queries and the prefix (length L) and appended L mask tokens as K and V.   

A second pass would include choosing a random integer 0\geq n <B for each sample where we start counting the blocks from.  

## Attention Changes:

The key difference from Eagle3 is that DFlash uses a non-causal $[L, 2L]$ attention pattern where queries attend to both the hidden states from the verifier (first $L$ positions) and the mask token embeddings (last $L$ positions), all in a single forward pass. 

## Model and Config Changes:

The DFlash model is based on the Qwen3 transformer architecture:
 
1.  It receives hidden state data from 5 layers of the network (hidden dimension $h$) - this forms the context, (similar to Eagle 3's 3 hidden states of data), and then takes in a block of mask tokens this forms the queries) of block size B. 
2. The $[5 \times h ,L]$ data goes through a fully connected layer which projects it down to $[h, L]$.  
3. The embedded mask is then appended to the context from the hidden states and passed into each of the 5 transformer layers - unmodified.  Layer 4 receives the same input for the K and V projections as layer 1, except for the very end of the sequences which comes from the prior layer's hidden states.  The embedded mask is used for query projections for the first layer, and the outputs for those queries are passed to the next layer, and so on.  
4. When inference is complete the model produces B tokens as the draft.  

The following changes are required to the existing model / config definitions:
- The DFlash model definition also needs a method to compute the loss and other metrics such as accuracy given a loss mask. There is currently a version of this for eagle 3 in models/eagle3/core.py
-  The model should be registered with the SpeculatorModel registry and work with the existing training infrastructure.
-  The model needs a configuration class with dflash specific parameters like 'block size'.  An example for eagle3 can be found in config.py
-  Generalize the trainer class in speculators/train/[trainer.py](http://trainer.py/) to accept DFlash configs and models.

# Implementation Steps:

- [ ] Update / add a dflash compatible training step which runs a single forward pass, as opposed to T steps. 
This forward pass and loss calculation should be performed in a file analogous to models/eagle3/[core.py](http://core.py/)
- [ ] Sequence trimming 
- [ ] Attention Updates:  Create the attention mask modifiers for flash attention. The attention mask within a document is demonstrated above, but we also need to mask out other documents when we concatenate the training data.
- [ ]  Model / Config Updates
  - [ ] Updated model definition
  - [ ] Update / add dflash specific parameters (e.g block_size) and reflect this value in the model's config.json
  - [ ] Generalize the trainer class in speculators/train/[trainer.py](http://trainer.py/) to accept DFlash configs and models.

**Comment by brian-dellabetta** [comment]
> The DFlash model is based on the Qwen3 transformer architecture

Is it safe to say that DFlash will always use Qwen3 similar to how it seems Eagle3 is in a sense hard-wired to use Llama? Just curious if this might be a blocker to adoption, as we've heard some orgs are not allowed to use Qwen
